library/modules/World.cpp

```diff
@@ -68,6 +75,133 @@ void World::SetPauseState(bool paused)
     DF_GLOBAL_VALUE(pause_state, dummy) = paused;
 }
 
+std::unordered_set<Lock*> PlayerLock::locks;
+std::unordered_set<Lock*> AnnouncementLock::locks;
+
+bool AnnouncementLock::isAnyLocked() const {
+    return std::any_of(locks.begin(), locks.end(), [](Lock* lock) { return lock->isLocked(); });
+}
+
+bool AnnouncementLock::isOnlyLocked() const {
+    return std::all_of(locks.begin(), locks.end(), [&](Lock* lock) {
+        if (lock == this) {
+            return lock->isLocked();
+        }
+        return !lock->isLocked();
+    });
+}
+
+bool PlayerLock::isAnyLocked() const {
+    return std::any_of(locks.begin(), locks.end(), [](Lock* lock) { return lock->isLocked(); });
+}
+
+bool PlayerLock::isOnlyLocked() const {
+    return std::all_of(locks.begin(), locks.end(), [&](Lock* lock) {
+        if (lock == this) {
+            return lock->isLocked();
+        }
+        return !lock->isLocked();
+    });
+}
```

Comment:
I would suggest moving the bodies of these functions into something that can be reused, to minimize duplication. (e.g. helper functions local to this file)

Reply:
I don't understand what you're getting at with this one. They are methods of the locks, making them re-usable?

Reply:
I was suffering code amnesia or something, I figured it out. It will be in 22cb7d20c8a0815fd46bd973f4e2b7d60b232e39